scanpy/tools/_rank_genes_groups.py

```diff
@@ -684,86 +683,118 @@ def filter_rank_genes_groups(
+            categories, obs_tidy = _prepare_dataframe(
+                adata, var_names, groupby='__is_in_cluster__', use_raw=use_raw,
+            )
 
-        # for if category defined by groupby (if any) compute for each var_name
-        # 1. the mean value over the category
-        # 2. the fraction of cells in the category having a value > 0
-
-        # 1. compute mean value
-        mean_obs = obs_tidy.groupby(level=0).mean()
-
-        # 2. compute fraction of cells having value >0
-        # transform obs_tidy into boolean matrix
-        obs_bool = obs_tidy.astype(bool)
-
-        # compute the sum per group which in the boolean matrix this is the number
-        # of values >0, and divide the result by the total number of values in the group
-        # (given by `count()`)
-        fraction_obs = obs_bool.groupby(level=0).sum() / obs_bool.groupby(level=0).count()
-
-        # Because the dataframe groupby is based on the '__is_in_cluster__' column,
-        # in this context, [True] means __is_in_cluster__.
-        # Also, in this context, fraction_obs.loc[True].values is the row of values
-        # that is assigned *as column* to fraction_in_cluster_matrix to follow the
-        # structure of the gene_names dataFrame
-        fraction_in_cluster_matrix.loc[:, cluster] = fraction_obs.loc[True].values
-        fraction_out_cluster_matrix.loc[:, cluster] = fraction_obs.loc[False].values
-
-        # compute fold change.
-        if log:
-            fold_change_matrix.loc[:, cluster] = (np.exp(mean_obs.loc[True]) / np.exp(mean_obs.loc[False])).values
-        else:
-            fold_change_matrix.loc[:, cluster] = (mean_obs.loc[True] / mean_obs.loc[False]).values
+            # for if category defined by groupby (if any) compute for each var_name
+            # 1. the fraction of cells in the category having a value > 0
+            # 2. the mean value over the category
+
+            # 1. compute fraction of cells having value >0
+            # transform obs_tidy into boolean matrix
+            obs_bool = obs_tidy.astype(bool)
+
+            # compute the sum per group which in the boolean matrix this is the number
+            # of values >0, and divide the result by the total number of values in the group
+            # (given by `count()`)
+            fraction_obs = (
+                obs_bool.groupby(level=0).sum() / obs_bool.groupby(level=0).count()
+            )
+
+            # Because the dataframe groupby is based on the '__is_in_cluster__' column,
+            # in this context, [True] means __is_in_cluster__.
+            # Also, in this context, fraction_obs.loc[True].values is the row of values
+            # that is assigned *as column* to fraction_in_cluster_matrix to follow the
+            # structure of the gene_names dataFrame
+            fraction_in_cluster_matrix.loc[:, cluster] = fraction_obs.loc[True].values
+            fraction_out_cluster_matrix.loc[:, cluster] = fraction_obs.loc[False].values
```

Comment:
Instead of making a huge dense array in `_prepare_dataframe`, could this look more like:

```suggestion
    def calc_frac_present(X):
        return np.ravel((X != 0).sum(axis=0)) / X.shape[0]

    for cluster in gene_names.columns:
        var_names = gene_names[cluster].values
        sub_adata = adata[:, var_names]
        sub_X = sc.get._get_obs_rep(sub_adata, use_raw=use_raw).copy()
        in_group = adata.obs[groupby] == cluster

        fraction_in_cluster_matrix.loc[:, cluster] = calc_frac_present(sub_X[in_group])
        fraction_out_cluster_matrix.loc[:, cluster] = calc_frac_present(sub_X[~in_group])
```

Reply:
Yes, ofc.